In [15]:
import geojson
import pandas as pd
import numpy as np

### District Shapes geojson file

In [1]:
pip install PyShp

Defaulting to user installation because normal site-packages is not writeable
     |████████████████████████████████| 46 kB 1.5 MB/s eta 0:00:011
You should consider upgrading via the '/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [16]:
import shapefile

shape = shapefile.Reader("nj_vest_20/nj_vest_20.shp")
#first feature of the shapefile
feature = shape.shapeRecords()[0]
first = feature.shape.__geo_interface__  
{'type': 'LineString', 'coordinates': ((0.0, 0.0), (25.0, 10.0), (50.0, 50.0))}


{'type': 'LineString', 'coordinates': ((0.0, 0.0), (25.0, 10.0), (50.0, 50.0))}

In [17]:
pip install dbfread

Defaulting to user installation because normal site-packages is not writeable
You should consider upgrading via the '/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [18]:
import csv
from dbfread import DBF

def dbf_to_csv(dbf_table_pth):#Input a dbf, output a csv, same name, same path, except extension
    csv_fn = dbf_table_pth[:-4]+ ".csv" #Set the csv file name
    table = DBF(dbf_table_pth)# table variable is a DBF object
    with open(csv_fn, 'w', newline = '') as f:# create a csv file, fill it with dbf content
        writer = csv.writer(f)
        writer.writerow(table.field_names)# write the column name
        for record in table:# write the rows
            writer.writerow(list(record.values()))
    return csv_fn# return the csv name


In [19]:
dbf_to_csv("nj_vtd_2020_bound.dbf")

'nj_vtd_2020_bound.csv'

In [20]:
with open('district-shapes.geojson') as f:
    gj = geojson.load(f)
features = gj['features'][0]
features

{"geometry": {"coordinates": [[[[-75.207824, 39.435387], [-75.207978, 39.436134], [-75.208899, 39.440615], [-75.209346, 39.442793], [-75.209499, 39.443546], [-75.209666, 39.44437], [-75.209791, 39.444976], [-75.210384, 39.448015], [-75.211421, 39.453359], [-75.211665, 39.45463], [-75.211878, 39.455728], [-75.213629, 39.455459], [-75.215733, 39.455138], [-75.218239, 39.452634], [-75.218929, 39.451955], [-75.219983, 39.451059], [-75.223203, 39.449631], [-75.224799, 39.448653], [-75.226119, 39.447311], [-75.227626, 39.446477], [-75.230541, 39.445819], [-75.232282, 39.44434], [-75.232957, 39.444331], [-75.235093, 39.445021], [-75.235258, 39.446779], [-75.236666, 39.446106], [-75.236725, 39.446114], [-75.239296, 39.446457], [-75.239791, 39.446524], [-75.242238, 39.446852], [-75.242851, 39.446934], [-75.244015, 39.44709], [-75.245549, 39.447295], [-75.246877, 39.447472], [-75.247571, 39.447383], [-75.248898, 39.446573], [-75.253019, 39.444056], [-75.252791, 39.44141], [-75.252713, 39.440367]

In [21]:
district_data = pd.json_normalize(gj['features'])
district_data

type geometry.type                               geometry.coordinates  \
0   Feature  MultiPolygon  [[[[-75.207824, 39.435387], [-75.207978, 39.43...   
1   Feature  MultiPolygon  [[[[-74.289202, 39.475292], [-74.296461, 39.47...   
2   Feature  MultiPolygon  [[[[-75.405489, 39.796368], [-75.42677, 39.788...   
3   Feature  MultiPolygon  [[[[-74.913152, 39.776754], [-74.913971, 39.77...   
4   Feature  MultiPolygon  [[[[-75.123821, 39.962744], [-75.125701, 39.96...   
5   Feature  MultiPolygon  [[[[-74.937525, 39.890294], [-74.938864, 39.89...   
6   Feature  MultiPolygon  [[[[-74.990068, 40.036435], [-75.000617, 40.02...   
7   Feature  MultiPolygon  [[[[-74.649348, 39.527271], [-74.647183, 39.52...   
8   Feature  MultiPolygon  [[[[-74.254198, 40.042967], [-74.253773, 40.04...   
9   Feature  MultiPolygon  [[[[-74.042222, 40.140702], [-74.042222, 40.14...   
10  Feature  MultiPolygon  [[[[-73.890791, 40.31069], [-73.971661, 40.322...   
11  Feature  MultiPolygon  [[[[-74.27624, 40.44412], [-74.276431, 40.4443...   
12  Feature  MultiPolygon  [[[[-74.033694, 40.375656], [-74.02858, 40.370...   
13  Feature  MultiPolygon  [[[[-74.629228, 40.338689], [-74.627877, 40.33...   
14  Feature  MultiPolygon  [[[[-74.771115, 40.214742], [-74.768574, 40.21...   
15  Feature  MultiPolygon  [[[[-74.752013, 40.431361], [-74.748147, 40.42...   
16  Feature  MultiPolygon  [[[[-74.428083, 40.523871], [-74.427131, 40.52...   
17  Feature  MultiPolygon  [[[[-74.337175, 40.585776], [-74.337236, 40.58...   
18  Feature  MultiPolygon  [[[[-74.242393, 40.599474], [-74.246079, 40.59...   
19  Feature  MultiPolygon  [[[[-74.260676, 40.661298], [-74.263726, 40.66...   
20  Feature  MultiPolygon  [[[[-74.498269, 40.563325], [-74.49729, 40.564...   
21  Feature  MultiPolygon  [[[[-74.456904, 40.5958], [-74.455132, 40.5959...   
22  Feature  MultiPolygon  [[[[-75.1896, 40.591776], [-75.189268, 40.5899...   
23  Feature  MultiPolygon  [[[[-74.914299, 40.858258], [-74.914274, 40.85...   
24  Feature  MultiPolygon  [[[[-74.344927, 41.192051], [-74.355228, 41.19...   
25  Feature  MultiPolygon  [[[[-74.351891, 41.045387], [-74.356897, 41.03...   
26  Feature  MultiPolygon  [[[[-74.179591, 40.887288], [-74.180613, 40.88...   
27  Feature  MultiPolygon  [[[[-74.262187, 40.715253], [-74.259903, 40.71...   
28  Feature  MultiPolygon  [[[[-74.187082, 40.723305], [-74.189217, 40.72...   
29  Feature  MultiPolygon  [[[[-74.042222, 40.140702], [-74.042286, 40.14...   
30  Feature  MultiPolygon  [[[[-74.157481, 40.766472], [-74.157688, 40.76...   
31  Feature  MultiPolygon  [[[[-74.03648, 40.752473], [-74.036533, 40.752...   
32  Feature  MultiPolygon  [[[[-74.088421, 40.769587], [-74.087879, 40.76...   
33  Feature  MultiPolygon  [[[[-74.193648, 40.761178], [-74.193373, 40.76...   
34  Feature  MultiPolygon  [[[[-74.167574, 40.935097], [-74.1681, 40.9360...   
35  Feature  MultiPolygon  [[[[-74.143385, 40.859397], [-74.143988, 40.85...   
36  Feature  MultiPolygon  [[[[-73.998404, 40.839824], [-73.996651, 40.83...   
37  Feature  MultiPolygon  [[[[-73.982475, 40.927702], [-73.982509, 40.92...   
38  Feature  MultiPolygon  [[[[-74.002801, 40.947235], [-74.002823, 40.94...   
39  Feature  MultiPolygon  [[[[-74.2804, 40.95999], [-74.280554, 40.95776...   

    properties.id properties.NAME  properties.TotalPop  properties.PopDevPct  \
0               1               1               229633             -0.010524   
1               2               2               232422              0.001493   
2               3               3               236205              0.017794   
3               4               4               233742              0.007181   
4               5               5               237187              0.022026   
5               6               6               234823              0.011839   
6               7               7               230455             -0.006982   
7               8               8               225591             -0.027941   
8     

In [22]:
## Let's check if the number of co-ordinates present is same as the number of precicncts in a district
district_data.loc[district_data['properties.id'] == 1,'geometry.coordinates'][0][0]

[[[-75.207824, 39.435387],
  [-75.207978, 39.436134],
  [-75.208899, 39.440615],
  [-75.209346, 39.442793],
  [-75.209499, 39.443546],
  [-75.209666, 39.44437],
  [-75.209791, 39.444976],
  [-75.210384, 39.448015],
  [-75.211421, 39.453359],
  [-75.211665, 39.45463],
  [-75.211878, 39.455728],
  [-75.213629, 39.455459],
  [-75.215733, 39.455138],
  [-75.218239, 39.452634],
  [-75.218929, 39.451955],
  [-75.219983, 39.451059],
  [-75.223203, 39.449631],
  [-75.224799, 39.448653],
  [-75.226119, 39.447311],
  [-75.227626, 39.446477],
  [-75.230541, 39.445819],
  [-75.232282, 39.44434],
  [-75.232957, 39.444331],
  [-75.235093, 39.445021],
  [-75.235258, 39.446779],
  [-75.236666, 39.446106],
  [-75.236725, 39.446114],
  [-75.239296, 39.446457],
  [-75.239791, 39.446524],
  [-75.242238, 39.446852],
  [-75.242851, 39.446934],
  [-75.244015, 39.44709],
  [-75.245549, 39.447295],
  [-75.246877, 39.447472],
  [-75.247571, 39.447383],
  [-75.248898, 39.446573],
  [-75.253019, 39.444056],
  [-7

### Precinct Assignment 

In [23]:
assign = pd.read_csv('precinct-assignments.csv')
assign.head(5)

GEOID20  District
0  34033020001         3
1  34033001501         3
2  34033042009         3
3  34033000703         3
4  34033042001         3

In [24]:
print(f"- We have {assign.shape[0]} number of precicncts assigned to {assign.District.nunique()} districts.\n- There are {np.count_nonzero(assign.isnull().values.ravel())} missing values in the dataset.\n- The statement all the precint ids in the dataset are unique is {assign.shape[0] == assign.GEOID20.nunique()}")

- We have 6754 number of precicncts assigned to 40 districts.
- There are 0 missing values in the dataset.
- The statement all the precint ids in the dataset are unique is True


In [25]:
assign.groupby(by="District").size()

District
1     214
2     125
3     171
4     154
5     179
6     213
7     213
8     182
9     180
10    170
11    166
12    152
13    175
14    156
15    150
16    178
17    148
18    182
19    169
20    163
21    198
22    195
23    193
24    192
25    182
26    165
27    172
28    220
29    238
30    103
31    195
32    166
33    137
34    156
35    110
36    117
37    130
38    145
39    141
40    159
dtype: int64

### Precinct Data

In [26]:
pre_data = pd.read_csv('precinct-data.csv')
pre_data.head(5)

GEOID20  District   FriendlyId  Total_2016-2021_Comp  \
0  34007043046         6  34007043046                   563   
1  34007175001         4  34007175001                   708   
2  34007040402         5  34007040402                   304   
3  34007135005         5  34007135005                   488   
4  34007005005         5  34007005005                   582   

   Dem_2016-2021_Comp  Rep_2016-2021_Comp  Total_2021_Gov  Dem_2021_Gov  \
0                 334                 220             433           265   
1                 251                 444             583           206   
2                 268                  30             131           106   
3                 310                 171             347           215   
4                 340                 223             434           238   

   Rep_2021_Gov  Total_2020_Sen  ...  Asian_2020_VAP.1  Native_2020_VAP.1  \
0           165             698  ...               184                 11   
1           374             866  ...                31                 12   
2            23             412  ...                 4                 23   
3           129             628  ...                45                 21   
4           193             722  ...                34                 13   

   Pacific_2020_VAP.1  Total_2020_TotalAdj  White_2020_TotalAdj  \
0                   6                 1568                  978   
1                   1                 1560                 1314   
2                   7                 1704                   42   
3                   4                 1118                  581   
4                   3                 1240                 1022   

   Hispanic_2020_TotalAdj  Black_2020_TotalAdj  Asian_2020_TotalAdj  \
0                     246                  104                  217   
1                     105                   71                   36   
2                    1232                  497                    6   
3                     290                  194                   66   
4                      71                   80                   49   

   Native_2020_TotalAdj  Pacific_2020_TotalAdj  
0                    18                      7  
1                    25                      1  
2                    41                      8  
3                    28                      4  
4                    22                      3  

[5 rows x 118 columns]

In [27]:
print(f"- We have {pre_data.shape[0]} number of precicncts assigned to {pre_data.District.nunique()} districts.\n- There are {np.count_nonzero(pre_data.isnull().values.ravel())} missing values in the dataset.\n- The statement all the precint ids in the dataset are unique is {pre_data.shape[0] == pre_data.GEOID20.nunique()}\n- The statement number of precicnts in assigned data and precinct data are same is {pre_data.GEOID20.nunique() == assign.GEOID20.nunique()}\n- The statement FriendlyID is the same as GEOID20 everywhere is {pre_data.loc[pre_data.GEOID20!=pre_data.FriendlyId].shape[0]==0}")

- We have 6383 number of precicncts assigned to 40 districts.
- There are 0 missing values in the dataset.
- The statement all the precint ids in the dataset are unique is True
- The statement number of precicnts in assigned data and precinct data are same is False
- The statement FriendlyID is the same as GEOID20 everywhere is False


### Look at the precinct data where GEOID20 is not the same as FriendlyID

In [28]:
pre_data.loc[pre_data.GEOID20!=pre_data.FriendlyId]

GEOID20  District            FriendlyId  \
6339      B_34005080001_AQAAAAA=         7   34005080001_part1_1   
6340      B_34005080001_/v///wM=         8  34005080001_part2_33   
6341          B_34007015005_/n8=         5  34007015005_part1_14   
6342          B_34007015005_AQA=         6   34007015005_part2_1   
6343  B_34007043008_AwAAAAAAAA==         5   34007043008_part1_2   
6344  B_34007043008_/P//////Aw==         6  34007043008_part2_48   
6345      B_34013070C23_AADgHw==        28   34013070C23_part1_8   
6346      B_34013070C23_//8fAA==        29  34013070C23_part2_21   
6347          B_34013070C28_AP4B        28   34013070C28_part1_8   
6348          B_34013070C28_/wEA        29   34013070C28_part2_9   
6349          B_34013070S01_Dw==        28   34013070S01_part1_4   
6350          B_34013070S01_8A==        29   34013070S01_part2_4   
6351          B_34013070S03_/4c/        28  34013070S03_part1_18   
6352          B_34013070S03_AHgA        29   34013070S03_part2_4   
6353      B_34013070S47_AQAAAAA=        28   34013070S47_part1_1   
6354      B_34013070S47_/v///38=        29  34013070S47_part2_38   
6355  B_34013070S49_//8AAAAAAA==        28  34013070S49_part1_16   
6356  B_34013070S49_AAD/////Pw==        29  34013070S49_part2_38   
6357          B_34013070W19_ADg=        28   34013070W19_part1_3   
6358          B_34013070W19_/wc=        29  34013070W19_part2_11   
6359          B_34013070W20_+P8=        28  34013070W20_part1_13   
6360          B_34013070W20_BwA=        29   34013070W20_part2_3   
6361          B_34013070W49_/A8=        28  34013070W49_part1_10   
6362          B_34013070W49_AwA=        29   34013070W49_part2_2   
6363          B_34017030B23_Ag==        31   34017030B23_part1_1   
6364          B_34017030B23_AQ==        32   34017030B23_part2_1   
6365          B_34017030B24_Pg==        31   34017030B24_part1_5   
6366          B_34017030B24_AQ==        32   34017030B24_part2_1   
6367          B_34017030B26_8P8=        31  34017030B26_part1_12   
6368          B_34017030B26_DwA=        32   34017030B26_part2_4   
6369          B_34017030E02_AwA=        31   34017030E02_part1_2   
6370          B_34017030E02_/AE=        32   34017030E02_part2_7   
6371          B_34017030E04_QA==        31   34017030E04_part1_1   
6372          B_34017030E04_Pw==        32   34017030E04_part2_6   
6373          B_34017030E06_CTg=        31   34017030E06_part1_5   
6374          B_34017030E06_9gc=        32   34017030E06_part2_9   
6375          B_34017030F06_iA==        31   34017030F06_part1_2   
6376          B_34017030F06_dw==        32   34017030F06_part2_6   
6377          B_34017030F08_Hg==        31   34017030F08_part1_4   
6378          B_34017030F08_AQ==        32   34017030F08_part2_1   
6379          B_34025ZZZZZZ_gAE=        11   34025ZZZZZZ_part1_2   
6380          B_34025ZZZZZZ_fwI=        13   34025ZZZZZZ_part2_8   
6381      B_34039040004_////Bw==        20  34039040004_part1_27   
6382      B_34039040004_AAAACA==        22   34039040004_part2_1   

      Total_2016-2021_Comp  Dem_2016-2021_Comp  Rep_2016-2021_Comp  \
6339                     0                   0                   0   
6340                   638                 311                 316   
6341                   393                 238                 147   
6342                     5                   5                   0   
6343                    26                  18                   6   
6344                   657                 473                 171   
6345                   166                 161                   5   
6346                   633                 611                  16   
6347                   149                 146                   3   
6348                   191                 185                   4   
6349                   112                 110                   2   
6350                   167                 161                   4   
6351                   467                 445        

In [29]:
pre_data.loc[pre_data.GEOID20!=pre_data.FriendlyId,'GEOID20'] = pre_data.loc[pre_data.GEOID20!=pre_data.FriendlyId, 'FriendlyId'].str.split("_").str[0]

In [30]:
pre_data.loc[pre_data.GEOID20!=pre_data.FriendlyId].tail(5)

GEOID20  District            FriendlyId  Total_2016-2021_Comp  \
6378  34017030F08        32   34017030F08_part2_1                    59   
6379  34025ZZZZZZ        11   34025ZZZZZZ_part1_2                     0   
6380  34025ZZZZZZ        13   34025ZZZZZZ_part2_8                     0   
6381  34039040004        20  34039040004_part1_27                   566   
6382  34039040004        22   34039040004_part2_1                    44   

      Dem_2016-2021_Comp  Rep_2016-2021_Comp  Total_2021_Gov  Dem_2021_Gov  \
6378                  54                   4              40            36   
6379                   0                   0               0             0   
6380                   0                   0               0             0   
6381                 215                 340             410           142   
6382                  24                  18              28            14   

      Rep_2021_Gov  Total_2020_Sen  ...  Asian_2020_VAP.1  Native_2020_VAP.1  \
6378             4              74  ...                66                  6   
6379             0               0  ...                 0                  0   
6380             0               0  ...                 0                  0   
6381           265             718  ...                67                 10   
6382            14              57  ...                 3                  0   

      Pacific_2020_VAP.1  Total_2020_TotalAdj  White_2020_TotalAdj  \
6378                   2                  291                   83   
6379                   0                    0                    0   
6380                   0                    0                    0   
6381                   0                 1298                  879   
6382                   0                  119                   62   

      Hispanic_2020_TotalAdj  Black_2020_TotalAdj  Asian_2020_TotalAdj  \
6378                      65                   71                   78   
6379                       0                    0                    0   
6380                       0                    0                    0   
6381                     241                   50                   97   
6382                      34                   20                    3   

      Native_2020_TotalAdj  Pacific_2020_TotalAdj  
6378                     9                      2  
6379                     0                      0  
6380                     0                      0  
6381                    16                      1  
6382                     0                      0  

[5 rows x 118 columns]

- Still not clean as there are some IDs with character literals in them

- Is there any relationship with the district?

In [31]:
pre_data.loc[pre_data.GEOID20!=pre_data.FriendlyId].groupby(by = "District").size()

District
5     2
6     2
7     1
8     1
11    1
13    1
20    1
22    1
28    9
29    9
31    8
32    8
dtype: int64

- Look at the precinct assignments, the indexes from 6341 to 6755 have GEOID of 15 characters instead of 11 characters.

### Boundary Data

In [32]:
bound_data = pd.read_csv("nj_vtd_2020_bound.csv")

In [35]:
bound_data.GEOID20.nunique() == bound_data.shape[0] #ALl unique values

True

In [56]:
x = assign.loc[assign.GEOID20.isin(assign_no_boundary)]

In [58]:
x.to_csv('precicnts_with_missing_data.csv', index=False)

In [45]:
pd.DataFrame(list(set(assign.GEOID20).difference(set(bound_data.GEOID20)))).to_csv("test_disparity.csv")

In [52]:
assign_no_boundary = list(set(assign.GEOID20).difference(set(bound_data.GEOID20)))
print(f"-There are {len(assign_no_boundary)} precincts that are assigned but don't have any boundary \n-Would be looking next to see if ther's any relationship with district")

-There are 415 precincts that are assigned but don't have any boundary 
-Would be looking next to see if ther's any relationship with district


In [41]:
len(set(assign.GEOID20))

6754

In [48]:
list(set(assign_no_data).difference(set(assign_no_boundary)))

[]

In [42]:
6754-6362

392

In [49]:
bound_data

STATEFP20  COUNTYFP20 VTDST20      GEOID20 VTDI20  \
0            34          35  005008  34035005008      P   
1            34          35  040212  34035040212      A   
2            34          35  040438  34035040438      A   
3            34          35  105001  34035105001      A   
4            34          35  030036  34035030036      A   
...         ...         ...     ...          ...    ...   
6356         34          33  042007  34033042007      P   
6357         34          33  60W001  3403360W001      A   
6358         34          33  050003  34033050003      A   
6359         34          33  060E01  34033060E01      A   
6360         34          33  004202  34033004202      A   

                                           NAME20  \
0           Bedminster township voting district 8   
1     Franklin township ward 2 voting district 12   
2     Franklin township ward 4 voting district 38   
3              Watchung borough voting district 1   
4         Bridgewater township voting district 36   
...                                           ...   
6356        Pennsville township voting district 7   
6357       Salem city West Ward voting district 1   
6358        Pittsgrove township voting district 3   
6359       Salem city East Ward voting district 1   
6360        Pennsville township voting district 2   

                                       NAMELSAD20  LSAD20 MTFCC20 FUNCSTAT20  \
0           Bedminster township voting district 8       0   G5240          S   
1     Franklin township ward 2 voting district 12       0   G5240          N   
2     Franklin township ward 4 voting district 38       0   G5240          N   
3              Watchung borough voting district 1       0   G5240          N   
4         Bridgewater township voting district 36       0   G5240          N   
...                                           ...     ...     ...        ...   
6356        Pennsville township voting district 7       0   G5240          S   
6357       Salem city West Ward voting district 1       0   G5240          N   
6358        Pittsgrove township voting district 3       0   G5240          N   
6359       Salem city East Ward voting district 1       0   G5240          N   
6360        Pennsville township voting district 2       0   G5240          N   

       ALAND20  AWATER20  INTPTLAT20  INTPTLON20  
0       563225         0   40.653805  -74.629211  
1      7875871     83176   40.533099  -74.545925  
2       964234         0   40.490488  -74.496713  
3      2144555     53833   40.636042  -74.461898  
4       757901         0   40.593669  -74.669536  
...        ...       ...         ...         ...  
6356    960989         0   39.643492  -75.524585  
6357   1916143   1083657   39.573320  -75.478334  
6358  25207861    598940   39.547879  -75.094973  
6359   2644376     74171   39.559004  -75.469171  
6360   1348264     11837   39.666667  -75.505787  

[6361 rows x 14 columns]

In [37]:
### Boundary and district

In [39]:
data_no_boundary = list(set(pre_data.GEOID20).difference(set(bound_data.GEOID20)))
print(f"-There are {len(data_no_boundary)} precincts that are assigned but don't have any boundary \n")

-There are 0 precincts that are assigned but don't have any boundary 



In [40]:
boundary_no_data = list(set(bound_data.GEOID20).difference(set(pre_data.GEOID20)))
print(f"-There are {len(boundary_no_data)} precincts that have boundaries but we have no corresponding data \n")

-There are 0 precincts that have boundaries but we have no corresponding data 



### Disparity between GEOID20 between two datasets

In [46]:
assign_no_data = list(set(assign.GEOID20).difference(set(pre_data.GEOID20)))
print(f"-There are {len(assign_no_data)} precincts that are assigned but don't have any data \n-Would be looking next to see if ther's any relationship with district")

-There are 415 precincts that are assigned but don't have any data 
-Would be looking next to see if ther's any relationship with district


In [13]:
assign.loc[assign.GEOID20.isin(assign_no_data)]

GEOID20  District
6339  340076066001015         5
6340  340076066001016         5
6341  340076066001020         5
6342  340076066001021         5
6343  340076066001022         5
...               ...       ...
6749  340170064001001        32
6750  340170064001002        32
6751  340170064001003        32
6752  340170064001004        32
6753  340170064001005        32

[415 rows x 2 columns]

In [14]:
index_ids = assign.loc[assign.GEOID20.isin(assign_no_data)].index.to_list()

In [15]:
index_ids == [i for i in range(6339,6754,1)]

True

- The reason we have missing data is because precinct assignments, the indexes from 6341 to 6755 have GEOID of 15 characters instead of 11 characters.